tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/CodeownersTools/CodeownersTools.cs

```diff
@@ -0,0 +1,445 @@
+            string repo,
+            string typeSpecProjectRoot,
+            string path = null,
+            string serviceLabel = null,
+            List<string> serviceOwners = null,
+            List<string> sourceOwners = null,
+            bool isAdding = false,
+            string workingBranch = null)
+        {
+            try
+            {
+                // Validate atleast Service Label or Path.
+                if (string.IsNullOrWhiteSpace(serviceLabel) && string.IsNullOrWhiteSpace(path))
+                {
+                    throw new Exception($"Service label: {serviceLabel} and Path: {path} are both invalid. At least one must be valid");
+                }
+
+                // Normalize service path
+                var normalizedPath = codeownersHelper.NormalizePath(path);
+
+                // Get labels file contents.
+                var labelsFileContent = await githubService.GetContentsSingleAsync(Constants.AZURE_OWNER_PATH, Constants.AZURE_SDK_TOOLS_PATH, Constants.AZURE_COMMON_LABELS_PATH);
+
+                if (labelsFileContent == null)
+                {
+                    throw new Exception("Could not retrieve labels file from the repository.");
+                }
+
+                var labelsContent = labelsFileContent.Content;
+
+                // Validate service label
+                if (!string.IsNullOrEmpty(serviceLabel))
+                {
+                    var serviceLabelValidationResults = LabelHelper.CheckServiceLabel(labelsContent, serviceLabel);
+                    if (serviceLabelValidationResults != LabelHelper.ServiceLabelStatus.Exists)
+                    {
+                        var labelsPullRequests = await githubService.SearchPullRequestsByTitleAsync(Constants.AZURE_OWNER_PATH, Constants.AZURE_SDK_TOOLS_PATH, "Service Label");
+
+                        if (!LabelHelper.CheckServiceLabelInReview(labelsPullRequests, serviceLabel) && string.IsNullOrEmpty(normalizedPath))
+                        {
+                            throw new Exception($"Service label: {serviceLabel} doesn't exist.");
+                        }
+                    }
+                }
+
+                if (string.IsNullOrEmpty(workingBranch))
+                {
+                    var codeownersPullRequests = await githubService.SearchPullRequestsByTitleAsync(Constants.AZURE_OWNER_PATH, repo, "CODEOWNERS");
+
+                    foreach (var codeownersPullRequest in codeownersPullRequests)
+                    {
+                        if (codeownersPullRequest != null &&
+                            ((!string.IsNullOrEmpty(serviceLabel) && codeownersPullRequest.Title.Contains(serviceLabel, StringComparison.OrdinalIgnoreCase)) ||
+                            (!string.IsNullOrEmpty(normalizedPath) && codeownersPullRequest.Title.Contains(normalizedPath, StringComparison.OrdinalIgnoreCase))))
+                        {
+                            workingBranch = codeownersPullRequest.Head.Ref;
```

Comment:
Typically after setting a value, you'd `break` out of the loop. Otherwise `workingBranch` keeps getting set to whatever PRs match the given criteria until the loop ends. How are `codeownersPullRequests` sorted? If you want the last one then it might make sense to traverse the list in reverse.